pkg/logql/syntax/ast.go

```diff
@@ -909,12 +909,19 @@ func (e *VectorAggregationExpr) String() string {
 // impl SampleExpr
 func (e *VectorAggregationExpr) Shardable() bool {
 	if e.Operation == OpTypeCount || e.Operation == OpTypeAvg {
-		// count is shardable is labels are not mutated
+		if !e.Left.Shardable() {
+			return false
+		}
+		// count is shardable if labels are not mutated
 		// otherwise distinct values can be counted twice per shard
 		shardable := true
-		e.Walk(func(e interface{}) {
+		e.Left.Walk(func(e interface{}) {
```

Comment:
It won't actually matter in this case (`e.Left.Walk` only walks over the child nodes as we already know what the current node is). In an earlier revision I was playing with it caused an infinite loop as I called `e.Shardable()` within the walk function -- not a good idea :sweat_smile: . I ended up putting the `e.Left.Shardable()` check outside the walk function because we only need call it once there rather than once for each child node recursively.